### PR TITLE
Clean smart ops

### PR DESCRIPTION
**Pull request by mayeul-zama**

### CODE REVIEW

File: tfhe/src/js_on_wasm_api/mod.rs [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
-pub mod boolean;
+mod boolean;
 #[cfg(feature = "boolean-client-js-wasm-api")]
 pub use boolean::*;
 
 #[cfg(feature = "parallel-wasm-api")]
 pub use wasm_bindgen_rayon::init_thread_pool;
 
-pub mod high_level_api;
-pub use high_level_api::*;
```
Comment (tmontaigu) [inline tfhe/src/js_on_wasm_api/mod.rs:15]: Why was it renamed ?

File: tfhe/src/shortint/engine/server_side/bitwise_op.rs [hunk 2]
```diff
@@ -29,14 +29,20 @@ impl ShortintEngine {
         Ok(())
     }
 
+    // by convention smart operations take mut refs to their inputs, even if they do not modify them
+    #[allow(clippy::needless_pass_by_ref_mut)]
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/bitwise_op.rs:33]: that #[allow] should not be needed anymore,  in the function, message_extract_assign was added, which makes &mut ct_left and &mut ct_right needed

File: tfhe/src/shortint/engine/server_side/bitwise_op.rs [hunk 3]
```diff
@@ -81,14 +87,20 @@ impl ShortintEngine {
...
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/bitwise_op.rs:91]: Same comment

File: tfhe/src/shortint/engine/server_side/bitwise_op.rs [hunk 4]
```diff
@@ -133,14 +145,20 @@ impl ShortintEngine {
...
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/bitwise_op.rs:149]: Same comment

File: tfhe/src/shortint/engine/server_side/comp_op.rs [hunk 5]
```diff
@@ -28,30 +28,21 @@ impl ShortintEngine {
         Ok(())
     }
 
+    // by convention smart operations take mut refs to their inputs, even if they do not modify them
+    #[allow(clippy::needless_pass_by_ref_mut)]
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/comp_op.rs:32]: same comment

File: tfhe/src/shortint/engine/server_side/comp_op.rs [hunk 6]
```diff
@@ -80,29 +71,21 @@ impl ShortintEngine {
...
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/comp_op.rs:75]: same comment

File: tfhe/src/shortint/engine/server_side/comp_op.rs [hunk 7]
```diff
@@ -131,29 +114,21 @@ impl ShortintEngine {
...
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/comp_op.rs:118]: Same comment

File: tfhe/src/shortint/engine/server_side/comp_op.rs [hunk 8]
```diff
@@ -182,29 +157,21 @@ impl ShortintEngine {
...
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/comp_op.rs:161]: Same comment

File: tfhe/src/shortint/engine/server_side/comp_op.rs [hunk 9]
```diff
@@ -233,35 +200,29 @@ impl ShortintEngine {
...
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/comp_op.rs:204]: Same Comment

File: tfhe/src/shortint/engine/server_side/comp_op.rs [hunk 10]
```diff
@@ -233,35 +200,29 @@ impl ShortintEngine {
...
         }
-        self.unchecked_equal_assign(server_key, ct_left, ct_right)?;
-        Ok(())
+        let mut result = ct_left.clone();
+        self.unchecked_equal_assign(server_key, &mut result, ct_right)?;
+        Ok(result)
     }
 
+    // by convention smart operations take mut refs to their inputs, even if they do not modify them
+    #[allow(clippy::needless_pass_by_ref_mut)]
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/comp_op.rs:221]: Same comment

File: tfhe/src/shortint/engine/server_side/comp_op.rs [hunk 11]
```diff
@@ -309,35 +270,29 @@ impl ShortintEngine {
         Ok(())
     }
 
+    // by convention smart operations take mut refs to their inputs, even if they do not modify them
+    #[allow(clippy::needless_pass_by_ref_mut)]
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/comp_op.rs:274]: Same comment

File: tfhe/src/shortint/engine/server_side/div_mod.rs [hunk 12]
```diff
@@ -39,14 +39,27 @@ impl ShortintEngine {
         Ok(())
     }
 
+    // by convention smart operations take mut refs to their inputs, even if they do not modify them
+    #[allow(clippy::needless_pass_by_ref_mut)]
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/div_mod.rs:43]: remove #[allow]

File: tfhe/src/shortint/engine/server_side/mod.rs [hunk 13]
```diff
@@ -593,6 +593,8 @@ impl ShortintEngine {
         Ok(())
     }
 
+    // by convention smart operations take mut refs to their inputs, even if they do not modify them
+    #[allow(clippy::needless_pass_by_ref_mut)]
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/mod.rs:597]: remove #[allow]

File: tfhe/src/shortint/engine/server_side/mul.rs [hunk 14]
```diff
@@ -166,14 +166,47 @@ impl ShortintEngine {
         Ok(())
     }
 
+    // by convention smart operations take mut refs to their inputs, even if they do not modify them
+    #[allow(clippy::needless_pass_by_ref_mut)]
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/mul.rs:170]: remove #[allow]

File: tfhe/src/shortint/engine/server_side/mul.rs [hunk 15]
```diff
@@ -187,16 +220,24 @@ impl ShortintEngine {
...
             self.message_extract_assign(server_key, ct_right)?;
         }
+
+        assert!(server_key.is_mul_possible(ct_left, ct_right));
         self.unchecked_mul_msb_assign(server_key, ct_left, ct_right)?;
         Ok(())
     }
 
+    // by convention smart operations take mut refs to their inputs, even if they do not modify them
+    #[allow(clippy::needless_pass_by_ref_mut)]
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/mul.rs:230]: remove #[allow]

File: tfhe/src/shortint/engine/server_side/scalar_add.rs [hunk 16]
```diff
@@ -45,12 +45,17 @@ impl ShortintEngine {
         Ok(())
     }
 
+    // by convention smart operations take mut refs to their inputs, even if they do not modify them
+    #[allow(clippy::needless_pass_by_ref_mut)]
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/scalar_add.rs:49]: remove #[allow]

File: tfhe/src/shortint/engine/server_side/scalar_mul.rs [hunk 17]
```diff
@@ -41,12 +41,18 @@ impl ShortintEngine {
         Ok(())
     }
 
+    // by convention smart operations take mut refs to their inputs, even if they do not modify them
+    #[allow(clippy::needless_pass_by_ref_mut)]
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/scalar_mul.rs:45]: remove #[allow]

File: tfhe/src/shortint/engine/server_side/scalar_sub.rs [hunk 18]
```diff
@@ -31,12 +31,17 @@ impl ShortintEngine {
         Ok(())
     }
 
+    // by convention smart operations take mut refs to their inputs, even if they do not modify them
+    #[allow(clippy::needless_pass_by_ref_mut)]
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/scalar_sub.rs:35]: remove #[allow]

File: tfhe/src/shortint/engine/server_side/shift.rs [hunk 19]
```diff
@@ -47,12 +47,17 @@ impl ShortintEngine {
         Ok(())
     }
 
+    // by convention smart operations take mut refs to their inputs, even if they do not modify them
+    #[allow(clippy::needless_pass_by_ref_mut)]
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/shift.rs:51]: remove #[allow]

File: tfhe/src/js_on_wasm_api/mod.rs (hunk 1 above)
Comment (mayeul-zama) [reply tfhe/src/js_on_wasm_api/mod.rs:15]: I took it from Arthur PR https://github.com/zama-ai/tfhe-rs/pull/470
It's already been merged in https://github.com/zama-ai/tfhe-rs/pull/494
I need to rebase this PR

Comment (tmontaigu) [reply tfhe/src/js_on_wasm_api/mod.rs]: Ok

**Comment by mayeul-zama** [comment]
Split into https://github.com/zama-ai/tfhe-rs/pull/535, https://github.com/zama-ai/tfhe-rs/pull/536, https://github.com/zama-ai/tfhe-rs/pull/537, https://github.com/zama-ai/tfhe-rs/pull/538 and https://github.com/zama-ai/tfhe-rs/pull/539

**Pull request closed without merging**